include/nbl/video/IGPUDescriptorSet.h

```diff
@@ -28,10 +30,131 @@ class NBL_API IGPUDescriptorSet : public asset::IDescriptorSet<const IGPUDescrip
+                // won't have uninitialized memory and destruction wouldn't crash in ~IGPUDescriptorSet.
+                std::uninitialized_default_construct_n(getDescriptorStorage(type) + m_descriptorStorageOffsets.data[i], m_layout->getTotalDescriptorCount(type));
+            }
+
+            const auto mutableSamplerCount = m_layout->getTotalMutableSamplerCount();
+            if (mutableSamplerCount > 0)
+                std::uninitialized_default_construct_n(getMutableSamplerStorage() + m_descriptorStorageOffsets.data[asset::EDT_COUNT], mutableSamplerCount);
+        }
+
+        inline uint64_t getVersion() const { return m_version.load(); }
+
+        inline void incrementVersion() { m_version.fetch_add(1ull); }
+
+        inline IDescriptorPool* getPool() const { return m_pool.get(); }
+
+        inline core::smart_refctd_ptr<asset::IDescriptor>* getAllDescriptors(const asset::E_DESCRIPTOR_TYPE type) const
+        {
+            auto* baseAddress = getDescriptorStorage(type);
+            if (baseAddress == nullptr)
+                return nullptr;
+
+            const auto offset = m_descriptorStorageOffsets.data[type];
+            if (offset == ~0u)
+                return nullptr;
+
+            return baseAddress + offset;
+        }
+
+        inline core::smart_refctd_ptr<IGPUSampler>* getAllMutableSamplers() const
+        {
+            auto* baseAddress = getMutableSamplerStorage();
+            if (baseAddress == nullptr)
+                return nullptr;
+
+            const auto poolOffset = m_descriptorStorageOffsets.data[asset::EDT_COUNT];
+            if (poolOffset == ~0u)
+                return nullptr;
+
+            return baseAddress + poolOffset;
+        }
+
+        // This assumes that descriptors of a particular type in the set will always be contiguous in pool's storage memory, regardless of which binding in the set they belong to.
+        inline core::smart_refctd_ptr<asset::IDescriptor>* getDescriptors(const asset::E_DESCRIPTOR_TYPE type, const uint32_t binding) const
+        {
+            const auto localOffset = getLayout()->getDescriptorOffset(type, binding);
+            if (localOffset == ~0)
+                return nullptr;
+
+            auto* descriptors = getAllDescriptors(type);
+            if (!descriptors)
+                return nullptr;
+
+            return descriptors + localOffset;
+        }
+
+        inline core::smart_refctd_ptr<IGPUSampler>* getMutableSamplers(const uint32_t binding) const
+        {
+            const auto localOffset = getLayout()->getMutableSamplerOffset(binding);
+            if (localOffset == ~0u)
+                return nullptr;
+
+            auto* samplers = getAllMutableSamplers();
+            if (!samplers)
+                return nullptr;
+
+            return samplers + localOffset;
+        }
```

Comment:
same as https://github.com/Devsh-Graphics-Programming/Nabla/pull/345/files#r1047774542

Reply:
basically nothing should be public

Reply:
Resolved.